core-bundle/src/Twig/Interop/ContaoEscaperNodeVisitor.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Twig\Interop;
+
+use Twig\Environment;
+use Twig\Node\Expression\ConstantExpression;
+use Twig\Node\Expression\FilterExpression;
+use Twig\Node\ModuleNode;
+use Twig\Node\Node;
+use Twig\NodeVisitor\AbstractNodeVisitor;
+
+/**
+ * This NodeVisitor alters all "escape('html')" filter expressions into
+ * "escape('contao_html')" filter expressions if the template they belong
+ * to is amongst the configured affected templates.
+ */
+class ContaoEscaperNodeVisitor extends AbstractNodeVisitor
+{
+    /**
+     * Evaluate affected templates on the fly so that they can be added after
+     * building the container.
+     *
+     * @var \Closure():array<string>
+     */
+    private $affectedTemplates;
+
+    /**
+     * @var array|null
+     */
+    private $escaperFilterNodes;
+
+    public function __construct(\Closure $affectedTemplates)
+    {
+        $this->affectedTemplates = $affectedTemplates;
+    }
+
+    /**
+     * Make sure to run after @see \Twig\NodeVisitor\EscaperNodeVisitor.
+     */
+    public function getPriority(): int
+    {
+        return 1;
+    }
+
+    protected function doEnterNode(Node $node, Environment $env): Node
+    {
+        if (
+            $node instanceof ModuleNode &&
+            \in_array($node->getTemplateName(), ($this->affectedTemplates)(), true)
+        ) {
+            $this->escaperFilterNodes = [];
+        } elseif (null !== $this->escaperFilterNodes && $this->isEscaperFilterExpressionWithHtmlStrategy($node)) {
+            $this->escaperFilterNodes[] = $node;
+        }
+
+        return $node;
+    }
+
+    protected function doLeaveNode(Node $node, Environment $env)
+    {
+        if ($node instanceof ModuleNode && null !== $this->escaperFilterNodes) {
+            foreach ($this->escaperFilterNodes as $escaperFilterNode) {
+                $this->setContaoEscaperArguments($escaperFilterNode);
+            }
+
+            $this->escaperFilterNodes = null;
+        }
+
+        return $node;
+    }
+
```

Comment:
The current implementation opens up a collection when an affected module is encountered, collects references to filter nodes and modifies them at once when leaving the module.

We could also toggle a flag when entering an affected module and directly modify each filter node when leaving it as long as the flag is set. Maybe this is cleaner because no references are needed. But I'm not sure.